### ISSUE TITLE
error C2248: 'Poco::Net::SocketImpl::init' : cannot access protected member declared in class 'Poco::Net::SocketImpl'

### ISSUE THREAD

**Issue by steji113** [pocoproject/poco#752]
I am getting a compilation error when trying to use a SecureSMTPClientSession. I am trying to update my codebase to use VS 2013, we used to use VS 2010. Because of this, I had upgraded my Poco version from 1.4.6p2-all (2013-10-22) to 1.6.0-all (2014-12-22).

This is possibly the offending code snippet that I am using:
`m_client.reset(new SecureSMTPClientSession(Host, static_cast<Poco::UInt16>(Port)));`

m_client is a unique_ptr.

**Comment by aleks-f** [comment]
All of POCO, including tests and samples, compiles with VS2013 without errors.

When you provide enough meaningful information about the said error and a line of code that is certainly (not possibly) offending, you may get a better answer.

**Comment by steji113** [comment]
I apologize for the lack of information. This is a Debug Win32 build.

So in the build output of VS 2013 I get this:

C:\Users\...\OutsideSources\Poco\Net\include\Poco/Net/Socket.h(626): error C2248: 'Poco::Net::SocketImpl::init' : cannot access protected member declared in class 'Poco::Net::SocketImpl'
3>          C:\Users\...\OutsideSources\Poco\Net\include\Poco/Net/SocketImpl.h(377) : see declaration of 'Poco::Net::SocketImpl::init'
3>          C:\Users\...\OutsideSources\Poco\Net\include\Poco/Net/SocketImpl.h(34) : see declaration of 'Poco::Net::SocketImpl'

Clicking the error actually brings me to the Socket.h class, which is why I am not exactly sure where the offending line is.

In my codebase I have SMTP utility class like so...
SMTPUtil.h:
```
#pragma once

#include "Poco\Net\SecureSMTPClientSession.h"

namespace net
{
namespace smtp
{
	class SMTPUtil
	{
	public:
		SMTPUtil();

		~SMTPUtil();

		bool Connect(const std::string &Host, const int Port, std::string &Error);

		bool Login(const std::string &Username, const std::string &Password, 
			std::string &Error);

		bool EnableTLS(std::string &Error);

		void Disconnect();

		bool SendMessage(const std::vector<std::string> &Recipients, const std::string &Subject,
			const std::string &Msg, const std::string &Sender,
			const std::string &SenderName, std::string &Error);

		bool SendMessage(const std::vector<std::string> &Recipients, const std::string &Subject,
			const std::string &Msg, const std::string &Sender,
			const std::string &SenderName, std::vector<BYTE> &Attachment,
			const std::string &AttachmentName, const std::string &AttachmentContentType,
			std::string &Error);
	private:
		std::unique_ptr<Poco::Net::SecureSMTPClientSession> m_client;
		bool m_connected;
	};
} // namespace smtp
} // namespace net
```

SMTPUtil.cpp
```
#include "stdafx.h"

#include "SMTPUtil.h"
#include "Poco\AutoPtr.h"

#pragma warning(disable : 4244)

#include "Poco\Net\MailMessage.h"
#include "Poco\Net\MailRecipient.h"
#include "Poco\Net\StringPartSource.h"
#include "Poco\Net\AcceptCertificateHandler.h"
#include "Poco\Net\InvalidCertificateHandler.h"
#include "Poco\Net\PrivateKeyPassphraseHandler.h"
#include "Poco\Net\KeyFileHandler.h"
#include "Poco\Net\SSLManager.h"
#include "Poco\Exception.h"
#include "Poco\Timespan.h"

using namespace Poco;
using namespace Poco::Net;
using namespace std;

namespace
{
	const long DEFAULT_READ_TIMEOUT_SEC = 10;
} // anonymouse namespace

namespace net
{
namespace smtp
{
	SMTPUtil::
	SMTPUtil()
	: m_connected(false)
	{
		// Do nothing
	}

	SMTPUtil::
	~SMTPUtil()
	{
		Disconnect();
	}

	bool
	SMTPUtil::
	Connect(const string &Host, const int Port, string &Error)
	{
		bool result = true;
		try {
			
			m_client.reset(new SecureSMTPClientSession(Host, static_cast<Poco::UInt16>(Port)));
			
			// Set receive timeout so socket operations don't wait forever
			Timespan timeout(DEFAULT_READ_TIMEOUT_SEC, 0);
			m_client->setTimeout(timeout);
			// Send EHLO or HELO only
			m_client->login();
			m_connected = true;
		}
		catch (Exception e)
		{
			result = false;
			Error = e.message();
		}

		return result;
	}

	bool
	SMTPUtil::
	Login(const string &Username, const string &Password, 
			string &Error)
	{
		if (!m_connected)
			return false;

		bool result = true;
		try {
			// Send EHLO or HELO and then login
			m_client->login(SMTPClientSession::AUTH_LOGIN, Username, Password);
		}
		catch (Exception e)
		{
			result = false;
			Error = e.message();
		}

		return result;
	}

	bool
	SMTPUtil::
	EnableTLS(string &Error)
	{
		if (!m_connected)
			return false;

		// Setup our context for verifying certs (currently don't care)
		SharedPtr<InvalidCertificateHandler> pHandler = new AcceptCertificateHandler(false);
		SharedPtr<PrivateKeyPassphraseHandler> pPassHandler = new KeyFileHandler(false);
		Context::Ptr pContext = new Context(Context::CLIENT_USE, "", "", "",
			Context::VERIFY_RELAXED, 9, true);
		pContext->enableExtendedCertificateVerification(false);

		// Init SSL client
		SSLManager::instance().initializeClient(pPassHandler, pHandler, pContext);

		bool started = false;
		try
		{
			started = m_client->startTLS(pContext);
		}
		catch (Exception e)
		{
			started = false;
			Error = e.message();
		}

		return started;
	}

	void
	SMTPUtil::
	Disconnect()
	{
		try
		{
			if (m_connected && m_client)
			{
				m_client->close();
				m_connected = false;
			}
		}
		catch (Exception e)
		{

		}
	}

	bool
	SMTPUtil::
	SendMessage(const vector<string> &Recipients, const string &Subject,
			const string &Msg, const std::string &Sender,
			const std::string &SenderName, string &Error)
	{
		if (!m_connected)
			return false;

		bool result = true;
		MailMessage message;
		// Add all recipients
		for(auto it = Recipients.begin(); it != Recipients.end(); ++it)
		{
			message.addRecipient(Poco::Net::MailRecipient(Poco::Net::MailRecipient::PRIMARY_RECIPIENT, *it));
		}
		// Set the subject, content and sender info
		message.setSubject(Subject);
		message.setContent(Msg);
		message.setSender(SenderName + " <" + Sender + ">");
		
		// Send the message
		try {
			m_client->sendMessage(message);
		}
		catch (Exception e)
		{
			result = false;
			Error = e.message();
		}

		return result;
	}

	bool
	SMTPUtil::
	SendMessage(const vector<string> &Recipients, const string &Subject,
			const string &Msg, const std::string &Sender,
			const std::string &SenderName, vector<BYTE> &Attachment, 
			const string &AttachmentName, const string &AttachmentContentType, string &Error)
	{
		if (!m_connected)
			return false;

		bool result = true;
		MailMessage message;
		// Add all recipients
		for(auto it = Recipients.begin(); it != Recipients.end(); ++it)
		{
			message.addRecipient(Poco::Net::MailRecipient(Poco::Net::MailRecipient::PRIMARY_RECIPIENT, *it));
		}
		// Set the subject, content and sender info
		message.setSubject(Subject);
		message.setSender(SenderName + " <" + Sender + ">");

		// Create a multipart entity containing the main message portion
		StringPartSource *pPartMsg = new StringPartSource(Msg);
		// NOTE: The message will takeover ownership of this part and destroy it
		message.addContent(pPartMsg);

		// Copy data into string to be able to create a multipart entity and
		// add to the message
		string data(Attachment.begin(), Attachment.end());
		StringPartSource *pPartAttach = new StringPartSource(data, AttachmentContentType);
		// NOTE: The message will takeover ownership of this part and destroy it
		message.addAttachment(AttachmentName, pPartAttach);

		// Send the message
		try {
			m_client->sendMessage(message);
		}
		catch (Exception e)
		{
			result = false;
			Error = e.message();
		}

		return result;
	}
} // namespace smtp
} // namespace net
```
If nothing looks awry here I can try to provide some more. Also, I will run the unit tests on my build just to double check.

**Comment by aleks-f** [comment]
Please, do not provide more. Help us help you - provide exactly a small code example that can be copied/pasted/compiled and demonstrates the problem. Nobody has time to read hundreds of lines of code and search for your problem for free. In other words, show that you have done some work yourself before asking other people for help.

**Comment by steji113** [comment]
So after much debugging, it turned out I had to forward declare the Socket class in Net/include/Poco/Net/SocketImpl.h like so:

```
namespace Poco {
namespace Net {
class Socket;
class Net_API SocketImpl: public Poco::RefCountedObject
```
It seems that the compiler options that are being used on my project are more strict than what is used to compile Poco w/ VS 2013.

**Comment by miguelangelorenes** [comment]
Hi,
I got the same problem and it was resolved like @steji113 tells. Why? I have my own "class Socket" and for any reason compiler is giving "friend grants" to my Socket class instead of Poco::Net::Socket class, so Including "class Socket" before SocketImpl declaration we are going to get the friend grants over SocketImpl are to Poco::Net::Socket instead of another class on your application environment(another third party library or your own code). 
I think that is an integration problem more than a Poco C++ issue.
Cheers!
